blobstore/cli/clustermgr/diskmgr.go

```diff
@@ -202,3 +224,121 @@ func openDiskTable(db *normaldb.NormalDB) (*normaldb.DiskTable, error) {
 	}
 	return tbl, nil
 }
+
+func cmdOfflineDisk(c *grumble.Context) error {
+	ctx := common.CmdContext()
+	cmClient := newCMClient(c.Flags)
+	diskid := args.DiskID(c.Args)
+	if diskid <= 0 {
+		return errors.New("invalid common args")
+	}
+
+	if !common.Confirm(fmt.Sprintf("offline disk %d?\n", diskid)) {
+		return nil
+	}
+
+	verbose := config.Verbose() || flags.Verbose(c.Flags)
+	if verbose {
+		fmt.Printf("set disk %d readonly\n", diskid)
+	}
+	if err := readonlyAndDropDisk(ctx, cmClient, diskid); err != nil {
+		return err
+	}
+	fmt.Printf("start to offline disk %d in the background\n", diskid)
+
+	return nil
+}
+
+func cmdOfflineAllDisks(c *grumble.Context) error {
+	ctx := common.CmdContext()
+	cmClient := newCMClient(c.Flags)
+	nodeHost := args.NodeHost(c.Args)
+	if len(nodeHost) <= 0 {
+		return errors.New("invalid args, node host is empty")
+	}
+
+	disks := make([]*blobnode.DiskInfo, 0)
+
+	// list all disks on the node
+	listOptionArgs := &clustermgr.ListOptionArgs{
+		Host:   nodeHost,
+		Status: proto.DiskStatusNormal, // only list normal disks
+		Marker: proto.DiskID(1),
+	}
+	for {
+		disksOneQuery, err := cmClient.ListDisk(ctx, listOptionArgs)
+		if err != nil {
+			return err
+		}
+		disks = append(disks, disksOneQuery.Disks...)
+
+		if disksOneQuery.Marker == proto.InvalidDiskID || len(disksOneQuery.Disks) < listOptionArgs.Count {
+			break
+		}
+	}
+
+	// show all disks on the node and confirm
+	verbose := config.Verbose() || flags.Verbose(c.Flags)
+	vv := flags.Vverbose(c.Flags)
+	fmt.Printf("disks on node <%s>:\n", nodeHost)
+	ac := common.NewAlternateColor(3)
+	for i, disk := range disks {
+		showDisk(disk, ac, i+1, verbose, vv)
+	}
+	if !common.Confirm(fmt.Sprintf("offline all disks on node %s?\n", nodeHost)) {
+		return nil
+	}
+
+	// offline disks concurrently
+	if verbose {
+		fmt.Printf("set all disks on node <%s> readonly\n", nodeHost)
+	}
+	var wg sync.WaitGroup
+	var errors []error
+	for _, disk := range disks {
+		diskid := disk.DiskID
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			if vv {
+				fmt.Printf("set disk %d readonly\n", diskid)
+			}
+			if err := readonlyAndDropDisk(ctx, cmClient, diskid); err != nil {
+				errors = append(errors, err)
```

Comment:
This line is not thread-safe. In my opinion, sequentially dropping disks is better than concurrently. Security of operation tools should be higher than performance requirements.